src/index.tsx

```diff
@@ -1039,20 +1039,36 @@ export function useFieldErrors<Value>(
 export function useFieldInitialValue<Value>(
   fieldAtom: FieldAtom<Value>,
   initialValue?: Value | typeof RESET,
-  options?: UseAtomOptions
+  options?: UseFieldInitialValueOptions<Value>
 ): UseFieldInitialValue {
   const field = useAtomValue(fieldAtom, options);
   const store = useStore(options);
+  const areEqual = (options && options.areEqual) || defaultValuesAreEqual;
 
-  React.useEffect(() => {
-    if (initialValue === undefined) {
-      return;
-    }
-    if (!store.get(field.dirty)) {
-      store.set(field.value, initialValue);
-    }
+  if (initialValue === undefined) {
+    return;
+  }
+
+  if (
+    !store.get(field.dirty) &&
+    !areEqual(initialValue, store.get(field.value))
+  ) {
+    store.set(field.value, initialValue);
+  }
+
+  if (!areEqual(initialValue, store.get(field._initialValue))) {
     store.set(field._initialValue, initialValue);
-  }, [store, field._initialValue, field.value, field.dirty, initialValue]);
+  }
+}
+
+function defaultValuesAreEqual(a: unknown, b: unknown): boolean {
+  if (Array.isArray(a) && Array.isArray(b)) {
+    return (
+      a.length === b.length && a.every((v, i) => defaultValuesAreEqual(v, b[i]))
+    );
+  }
+
+  return Object.is(a, b);
```

Comment:
we need to check first for object is, as with the initialValue, we will unnecessarily do deep array compare, while very often those will be the same array.

`object.is(a,b) || arraysAreEqual(a,b)`